crates/pera-core/src/signature_mpc/mpc_manager.rs

```diff
@@ -2,163 +2,231 @@ use crate::authority::authority_per_epoch_store::AuthorityPerEpochStore;
+            return None;
+        };
+        Some(ConsensusTransaction::new_signature_mpc_message(
+            epoch_store.name,
+            bcs::to_bytes(&message).unwrap(),
+            self.session_id.clone(),
+        ))
+    }
+
+    fn store_message(
+        &mut self,
+        message: &ProofMPCMessage,
+        epoch_store: Arc<AuthorityPerEpochStore>,
+    ) -> PeraResult<()> {
+        let party_id = authority_name_to_party_id(message.authority, &epoch_store)?;
+        if self.pending_messages.contains_key(&party_id) {
+            // TODO(#260): Punish an authority that sends multiple messages in the same round
+            return Ok(());
+        }
+
+        match bcs::from_bytes(&message.message) {
+            Ok(message) => {
+                self.pending_messages.insert(party_id, message);
+                Ok(())
             }
-        });
+            Err(err) => Err(PeraError::ObjectDeserializationError {
+                error: err.to_string(),
+            }),
+        }
     }
 
-    async fn handle_mpc_proof_init_event(
-        public_parameters: maurer::language::PublicParameters<
-            { maurer::SOUND_PROOFS_REPETITIONS },
-            Lang,
-        >,
-        consensus_adapter: Arc<dyn SubmitToConsensus>,
-        epoch_store: Weak<AuthorityPerEpochStore>,
-        threshold: usize,
+    async fn start_proof_mpc_flow(
+        &self,
+        public_parameters: ProofPublicParameters,
     ) -> anyhow::Result<ProofParty> {
         let batch_size = 1;
-        let party_state: ProofParty =
-            proof::aggregation::asynchronous::Party::new_proof_round_party(
-                public_parameters,
-                PhantomData,
-                threshold as group::PartyID,
-                batch_size,
-                &mut OsRng,
-            )?;
-
-        match party_state.advance(HashMap::new(), &(), &mut OsRng) {
-            Ok(advance_result) => {
-                let AdvanceResult::Advance((message, new_party)) = advance_result else {
-                    return Err(anyhow!("Finalization reached unexpectedly"));
-                };
-                let Some(epoch_store) = epoch_store.upgrade() else {
-                    // TODO: (#259) Handle the case when the epoch switched in the middle of the MPC instance
-                    return Err(anyhow!("Epoch store not found"));
-                };
-                let message_tx = ConsensusTransaction::new_signature_mpc_message(
-                    epoch_store.name,
-                    bcs::to_bytes(&message)?,
-                );
-                consensus_adapter
-                    .submit_to_consensus(&vec![message_tx], &epoch_store)
+        let party: ProofParty = proof::aggregation::asynchronous::Party::new_proof_round_party(
+            public_parameters,
+            PhantomData,
+            self.mpc_threshold_number_of_parties as PartyID,
```

Comment:
The `threshold` field being of type `PartyID` doesn't make any sense; @ycscaly it looks like this is comming from the
Maurer proof code.